File: day16/day16.ts
import { assert } from "../lib";
import { parseFile } from "../lib/fileParser";

const input = parseFile(__dirname);

function partOneSolution(input: string, phases: number): string {
  const basePattern = [0, 1, 0, -1];
  let currentInput = input;

  // For phases
  for (let i = 0; i < phases; i++) {
    let newInput = "";

    // For each cycle
    for (let j = 1; j <= currentInput.length; j++) {
      const pattern = getPatternForPosition(basePattern, j);

      let total = 0;

      // For each digig on a row
      for (let k = 0; k < currentInput.length; k++) {
        const digit = Number(currentInput[k]);
        const patternIndex = k % pattern.length;
        const patternNum = pattern[patternIndex];
        total += digit * patternNum;
      }

      newInput += Math.abs(Number(total.toString().slice(-1))).toString();
    }

    currentInput = newInput;
  }
  return currentInput.slice(0, 8);
}

/**
 * Helper methods
 */

function getPatternForPosition(pattern: number[], set: number): number[] {
  const newPattern = flatten(pattern.map(num => Array(set).fill(num)));

  let temp = newPattern[0];
  newPattern.splice(0, 1);
  newPattern.push(temp);
  return newPattern;
}

function flatten<T = any>(array: any): T[] {
  return array.reduce((acc: T[], item: any) => {
    return Array.isArray(item) ? [...acc, ...flatten(item)] : [...acc, item];
  }, []);
}

/**
 * Test
 */

const testInput = "12345678";

// assert(partOneSolution(testInput, 1), "48226158");
// assert(partOneSolution(testInput, 2), "34040438");
// assert(partOneSolution(testInput, 3), "03415518");
// assert(partOneSolution(testInput, 4), "01029498");

// assert(partOneSolution("80871224585914546619083218645595", 100), "24176176");

// console.log(1 * 1 + 2 * 0 + 3 * -1 + 4 * 0 + 5 * 1 + 6 * 0 + 7 * -1 + 8 * 0);
// console.log(1 * 0 + 2 * 1 + 3 * 1 + 4 * 0 + 5 * 0 + 6 * -1 + 7 * -1 + 8 * 0);

console.log(partOneSolution(input, 100));

File: day16/day16.go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	// input, err := ioutil.ReadFile("input.txt")

	// if err != nil {
	// 	fmt.Println("File reading error", err)
	// 	return
	// }

	// inputStr := string(input)
	// res1 := partOneSolution(inputStr, 100)

	base := [4]int{0, 1, 0, -1}
	pattern := getPatternForPosition(base, 2)
	fmt.Println(pattern)

	// fmt.Printf("Test 1: %s\n", partOneSolution("12345678", 1))
	// fmt.Printf("Test 1: %s\n", partOneSolution("12345678", 2))
	// fmt.Printf("Test 1: %s\n", partOneSolution("12345678", 3))
	// fmt.Printf("Test 1: %s\n", partOneSolution("12345678", 4))

	// fmt.Printf("Part One Solution: %s\n", res1) // 53296082
}

func partOneSolution(input string, phases int) string {
	basePattern := [4]int{0, 1, 0, -1}
	currentInput := input

	for i := 0; i < phases; i++ {
		newInput := ""

		for j := 1; j < len(currentInput); j++ {
			pattern := getPatternForPosition(basePattern, j)
			patternLength := len(pattern)
			total := 0

			for k, char := range currentInput {
				digit := int(char)
				patternNum := pattern[k%patternLength]
				total += digit * patternNum
			}

			newInput += strconv.Itoa(getLastNum(total))
		}
		fmt.Println(newInput)
		currentInput = newInput
	}
	return substring(currentInput, 0, 8)
}

// [ ,]

func getPatternForPosition(basePattern [4]int, set int) []int {
	basePatternLength := len(basePattern) * set
	result := make([]int, basePatternLength)
	currIndex := 0
	for j := range basePattern {
		for i := 0; i < set; i++ {
			if j == 3 && i == set-1 {
				break
			}

			if j+1 < 4 {
				result[currIndex] = basePattern[j+1]
			}
			currIndex++
		}
	}

	fmt.Println(result)
	result[basePatternLength-1] = basePattern[0]

	// tmp := result[0]
	// for i, n := range result {
	// 	if i == 0 {
	// 		continue
	// 	}

	// 	result[i-1] = n
	// }

	result[0] = basePattern[1]
	// result[basePatternLength-1] = tmp
	return result
}

// Get last number
func getLastNum(num int) int {
	if num < 0 {
		num *= -1
	}
	return num % 10
}

func substring(str string, start int, end int) string {
	runes := []rune(str)
	// ... Convert back into a string from rune slice.
	return string(runes[start:end])
}
